iceberg caching catalog refresh table failed

The issue seems to be introduced by https://github.com/StarRocks/starrocks/pull/53007

on table refresh we get this stacktrace
```
 [CachingIcebergCatalog.refreshCatalog():385] refresh landing_crm_tj.users metadata cache failed, msg :
com.google.common.util.concurrent.UncheckedExecutionException: java.lang.IllegalArgumentException: Cannot find snapshot with ID 6735346390613841142
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2085) ~[spark-dpp-1.0.0.jar:?]
        at com.google.common.cache.LocalCache.get(LocalCache.java:4011) ~[spark-dpp-1.0.0.jar:?]
        at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:4034) ~[spark-dpp-1.0.0.jar:?]
        at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:5010) ~[spark-dpp-1.0.0.jar:?]
        at com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(LocalCache.java:5017) ~[spark-dpp-1.0.0.jar:?]
        at com.starrocks.connector.iceberg.CachingIcebergCatalog.refreshTable(CachingIcebergCatalog.java:287) ~[starrocks-fe.jar:?]
        at com.starrocks.connector.iceberg.CachingIcebergCatalog.refreshTable(CachingIcebergCatalog.java:273) ~[starrocks-fe.jar:?]
        at com.starrocks.connector.iceberg.CachingIcebergCatalog.refreshCatalog(CachingIcebergCatalog.java:327) ~[starrocks-fe.jar:?]
        at com.starrocks.connector.hive.ConnectorTableMetadataProcessor.refreshIcebergCachingCatalog(ConnectorTableMetadataProcessor.java:146) ~[starrocks-fe.jar:?]
        at com.starrocks.connector.hive.ConnectorTableMetadataProcessor.runAfterCatalogReady(ConnectorTableMetadataProcessor.java:96) ~[starrocks-fe.jar:?]
        at com.starrocks.common.util.FrontendDaemon.runOneCycle(FrontendDaemon.java:72) ~[starrocks-fe.jar:?]
        at com.starrocks.common.util.Daemon.run(Daemon.java:98) ~[starrocks-fe.jar:?]
Caused by: java.lang.IllegalArgumentException: Cannot find snapshot with ID 6735346390613841142
        at org.apache.iceberg.relocated.com.google.common.base.Preconditions.checkArgument(Preconditions.java:204) ~[iceberg-bundled-guava-1.7.1.jar:?]
        at org.apache.iceberg.SnapshotScan.useSnapshot(SnapshotScan.java:85) ~[iceberg-core-1.7.1.jar:?]
        at org.apache.iceberg.BaseTableScan.useSnapshot(BaseTableScan.java:25) ~[iceberg-core-1.7.1.jar:?]
        at com.starrocks.connector.iceberg.IcebergCatalog.getPartitions(IcebergCatalog.java:197) ~[starrocks-fe.jar:?]
        at com.starrocks.connector.iceberg.CachingIcebergCatalog.lambda$new$0(CachingIcebergCatalog.java:101) ~[starrocks-fe.jar:?]
        at com.google.common.cache.CacheLoader$FunctionToCacheLoader.load(CacheLoader.java:169) ~[spark-dpp-1.0.0.jar:?]
        at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3570) ~[spark-dpp-1.0.0.jar:?]
        at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2312) ~[spark-dpp-1.0.0.jar:?]
        at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2189) ~[spark-dpp-1.0.0.jar:?]
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2079) ~[spark-dpp-1.0.0.jar:?]
        ... 11 more
```

`Cannot find snapshot with ID` is a message from iceberg 'SnapshotScan.useSnapshot'
i verified that snapshot do exist, but the error is real too

so the issue seems to be that we try to set this snapshot to metadata table scan, which is not the same as base table
https://github.com/StarRocks/starrocks/pull/53007#discussion_r1951505462

--- a/fe/fe-core/src/main/java/com/starrocks/connector/iceberg/CachingIcebergCatalog.java
+++ b/fe/fe-core/src/main/java/com/starrocks/connector/iceberg/CachingIcebergCatalog.java
@@ -271,28 +271,33 @@ public class CachingIcebergCatalog implements IcebergCatalog {
                 invalidateCache(icebergTableName);
                 return;
             }
+
             if (!currentLocation.equals(updateLocation)) {
                 LOG.info("Refresh iceberg caching catalog table {}.{} from {} to {}",
                         dbName, tableName, currentLocation, updateLocation);
-                long baseSnapshotId = currentOps.current().currentSnapshot().snapshotId();
-                refreshTable(updateTable, baseSnapshotId, dbName, tableName, executorService);
+                refreshTable(currentTable, updateTable, dbName, tableName, executorService);
                 LOG.info("Finished to refresh iceberg table {}.{}", dbName, tableName);
             }
         }
     }
 
-    private void refreshTable(BaseTable updatedTable, long baseSnapshotId,
+    private void refreshTable(BaseTable currentTable, BaseTable updatedTable,
                               String dbName, String tableName, ExecutorService executorService) {
+        long baseSnapshotId = currentTable.currentSnapshot().snapshotId();
         long updatedSnapshotId = updatedTable.currentSnapshot().snapshotId();
         IcebergTableName baseIcebergTableName = new IcebergTableName(dbName, tableName, baseSnapshotId);
         IcebergTableName updatedIcebergTableName = new IcebergTableName(dbName, tableName, updatedSnapshotId);
         long latestRefreshTime = tableLatestRefreshTime.computeIfAbsent(new IcebergTableName(dbName, tableName), ignore -> -1L);
 
-        partitionCache.invalidate(baseIcebergTableName);
-        partitionCache.getUnchecked(updatedIcebergTableName);
+        // update tables before refresh partition cache
+        // so when refreshing partition cache, `getTables` can return the latest one.
+        // another way to fix is to call `delegate.getTables` when refreshing partition cache.
         synchronized (this) {
             tables.put(updatedIcebergTableName, updatedTable);
         }
+
+        partitionCache.invalidate(baseIcebergTableName);
+        partitionCache.getUnchecked(updatedIcebergTableName);
 
         TableMetadata updatedTableMetadata = updatedTable.operations().current();
         List<ManifestFile> manifestFiles = updatedTable.currentSnapshot().dataManifests(updatedTable.io()).stream()

--- a/fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergCatalog.java
+++ b/fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergCatalog.java
@@ -14,6 +14,7 @@
 
 package com.starrocks.connector.iceberg;
 
+import com.google.common.base.Preconditions;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.Maps;
 import com.starrocks.catalog.Database;
@@ -244,7 +245,13 @@ public interface IcebergCatalog extends MemoryTrackable {
         PartitionsTable partitionsTable = (PartitionsTable) MetadataTableUtils.
                 createMetadataTableInstance(nativeTable, MetadataTableType.PARTITIONS);
         TableScan tableScan = partitionsTable.newScan();
+        // NOTE: if there is an exception raise because of snapshot id is not the latest one, it's expected
+        // using partition metadata table scan is more  efficient than doing file scan, but limitation is
+        // it only supports the latest snapshot id.
         if (snapshotId != -1) {
+            Preconditions.checkArgument(nativeTable.currentSnapshot().snapshotId() == snapshotId,
+                    "Ignore this error if snapshot id does not match. Iceberg partition metadata table only supports latest " +
+                            "snapshot. current = " + nativeTable.currentSnapshot().snapshotId() + ", expect = " + snapshotId);
             tableScan = tableScan.useSnapshot(snapshotId);
         }
         if (executorService != null) {

